fix: reverse the PREFIXes when adding a BASE to the prologue

sergiofenoll opened the pull request:
The AST of the SPARQL query stores the prefixes in reverse order for some reason. This isn't necessarily the right place to fix this, but it's an easy spot to reverse the reversal and ensure that the prefixes are following the same order that the user supplied them in.

The order of the prefixes is important because if two prefixes are defined, the latter "wins" and shadows the former. As of writing, the JS template automatically adds some prefixes, one of which is mu with an unexpected URI expansion. Without this fix, this ends up breaking queries that depend on the mu prefix.

A concrete example, the following query:

```sparql
PREFIX besluitvorming: <https://data.vlaanderen.be/ns/besluitvorming#>
PREFIX mu: <http://mu.semte.ch/vocabularies/core/>

PREFIX muExt: <http://mu.semte.ch/vocabularies/ext/>
PREFIX muCore: <http://mu.semte.ch/vocabularies/core/>
PREFIX mu: <http://mu.semte.ch/vocabularies/>
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT DISTINCT ?agendaitem ?newDocument
WHERE {
  ?newDocument mu:uuid """6634C7AFF96D38DDE006A4E7""" .
  ?agendaitem mu:uuid """f1e42414-0932-11ef-b935-71fbda946ed9""" .
}
```

Gets turned into:

```sparql
BASE <http://mu.semte.ch/prefix/local/> PREFIX besluitvorming: <https://data.vlaanderen.be/ns/besluitvorming#> PREFIX mu: <http://mu.semte.ch/vocabularies/core/> PREFIX muExt: <http://mu.semte.ch/vocabularies/ext/> PREFIX muCore: <http://mu.semte.ch/vocabularies/core/> PREFIX mu: <http://mu.semte.ch/vocabularies/> PREFIX xsd: <http://www.w3.org/2001/XMLSchema#> SELECT DISTINCT ?agendaitem ?newDocument FROM <http://mu.semte.ch/graphs/system/parliament> FROM <http://mu.semte.ch/graphs/system/signing> FROM <http://mu.semte.ch/graphs/system/email> FROM <http://mu.semte.ch/graphs/organizations/kanselarij> FROM <http://mu.semte.ch/graphs/system/users> FROM <http://mu.semte.ch/graphs/staatsblad> FROM <http://mu.semte.ch/graphs/sessions> FROM <http://mu.semte.ch/graphs/public> WHERE { ?newDocument mu:uuid """6634C7AFF96D38DDE006A4E7""" . ?agendaitem mu:uuid """f1e42414-0932-11ef-b935-71fbda946ed9""" . }
```

Notice that the last encountered `PREFIX mu:` is  `<http://mu.semte.ch/vocabularies/>`, which isn't the expected value.

After the fix it becomes:

```sparql
BASE <http://mu.semte.ch/prefix/local/> PREFIX xsd: <http://www.w3.org/2001/XMLSchema#> PREFIX mu: <http://mu.semte.ch/vocabularies/> PREFIX muCore: <http://mu.semte.ch/vocabularies/core/> PREFIX muExt: <http://mu.semte.ch/vocabularies/ext/> PREFIX mu: <http://mu.semte.ch/vocabularies/core/> PREFIX besluitvorming: <https://data.vlaanderen.be/ns/besluitvorming#> SELECT DISTINCT ?agendaitem ?newDocument FROM <http://mu.semte.ch/graphs/system/parliament> FROM <http://mu.semte.ch/graphs/system/signing> FROM <http://mu.semte.ch/graphs/system/email> FROM <http://mu.semte.ch/graphs/organizations/kanselarij> FROM <http://mu.semte.ch/graphs/system/users> FROM <http://mu.semte.ch/graphs/staatsblad> FROM <http://mu.semte.ch/graphs/sessions> FROM <http://mu.semte.ch/graphs/public> WHERE { ?newDocument mu:uuid """6634C7AFF96D38DDE006A4E7""" . ?agendaitem mu:uuid """f1e42414-0932-11ef-b935-71fbda946ed9""" . }
```

sergiofenoll committed:
fix: reverse the PREFIXes when adding a BASE to the prologue

The AST of the SPARQL query stores the prefixes in reverse order for
some reason. This isn't necessarily the right place to fix this, but
it's an easy spot to reverse the reversal and ensure that the prefixes
are following the same order that the user supplied them in.

The order of the prefixes is important because if two prefixes are
defined, the latter "wins" and shadows the former. As of writing, the JS
template automatically adds some prefixes, one of which is mu with an
unexpected URI expansion. Without this fix, this ends up breaking
queries that depend on the mu prefix.

madnificent commented:
Can you verify the example query input?  I think I understand what you want to express and that the requested query example should be the following instead:

``` sparql
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
PREFIX mu: <http://mu.semte.ch/vocabularies/>
PREFIX muCore: <http://mu.semte.ch/vocabularies/core/>
PREFIX muExt: <http://mu.semte.ch/vocabularies/ext/>

PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
PREFIX besluitvorming: <https://data.vlaanderen.be/ns/besluitvorming#>

SELECT DISTINCT ?agendaitem ?newDocument
WHERE {
  ?newDocument mu:uuid """6634C7AFF96D38DDE006A4E7""" .
  ?agendaitem mu:uuid """f1e42414-0932-11ef-b935-71fbda946ed9""" .
}
```

Can you confirm this is in a SELECT query and not the selection that happens for a MODIFY (`DELETE {} INSERT {} WHERE {}` or `DELETE {} WHERE {}` or `INSERT {} WHERE {}`)?

madnificent committed:
Only apply fix for prefixes in quad detection

The prefixes are stored in reverse order in the *info* object filled
when detecting quads.  The reverse order is faster and makes detection
of prefixes within the query faster.  However, the reverse order is
incorrect when generating the query for the WHERE clause.

An earlier fix reversed the order of the PREFIXES for all queries,
rather than only those of the update.  This code applies the change only
where necessary.

co-author: @erikap

madnificent commented:
We have altered the change in order to apply the fix only where we see it is currently necessary and will merge that into master.  This will automatically close this PR.  Feel free to re-open the PR if our assumptions were not correct.

madnificent committed:
Merge branch 'master' into fix/reverse-prefixes
